### PR TITLE
Add option to set the encoder A and B pins on the MCU to use internal pull-down, pull-up, or leave floating

### PR DESCRIPTION
**Pull request by Croktopus**
## Description

Added 3 toggles: ENC_PUP, ENC_PDOWN, and ENC_FLOAT. When ENC_PUP is defined, it sets the internal resistor on the encoder pins to be pulled up, and the same sort of behavior for the others. If none of these are defined, then it defaults to the current behavior of having the internal pull-up.

My optical encoder needs a pull down on these pins to work right, and the values also need to be finely tuned which might be possible even with an internal pull-up, but it's way easier if I can just set the pin to be floating.

I'm not much of a coder, so while I did test it and it works on my hardware, it looks super inelegant and I'm sure theres like a billion ways to optimize the code. I'm also not married to the names of those toggle things.

I also tried my hand at putting in some documentation based on existing docs.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/encoder.c [hunk 1]
```diff
@@ -81,12 +81,35 @@ void encoder_init(void) {
...
+
+        encoder_state[i] = (readPin(encoders_pad_a[i]) << 0) | (readPin(encoders_pad_b[i]) << 1);
+    }
+#else
+    for (int i = 0; i < NUMBER_OF_ENCODERS; i++) {
+        setPinInputHigh(encoders_pad_a[i]);
+        setPinInputHigh(encoders_pad_b[i]);
+
+        encoder_state[i] = (readPin(encoders_pad_a[i]) << 0) | (readPin(encoders_pad_b[i]) << 1);
+    }
```
Comment (drashna) [inline quantum/encoder.c:111]: This appears to just be a repeat of ENC_PUP, and can be removed. 
```suggestion
```

**Review by drashna — COMMENTED** [review]
It would be better to have something like `ENCODER_PIN_CFG` and have that set to the different types.  

Much like the col2row stuff.   This way, we can check to see if it's defined or not, and default to pullup. 

However, `ENCODER_PIN_CFG` is probably a horrible name, and something better should be used.

File: quantum/encoder.c [hunk 2]
```diff
@@ -81,12 +81,35 @@ void encoder_init(void) {
...
+#endif
```
Comment (mtei) [inline quantum/encoder.c:112]: For example, the following makes your code simple and easy to maintain.

Add the following code to `quantum/config_common.h`. 
```c
/* encoder pin mode */
#define ENC_PUP   0
#define ENC_PDOWN 1
#define ENC_FLOAT 2
```

The user configures the following in `config.h`: 
```c
#define ENCODER_PIN_MODE ENC_PUP /* or ENC_PDOWN, ENC_FLOAT */
```

```suggestion
#ifdef ENCODER_PIN_MODE
#  if ENCODER_PIN_MODE == ENC_PUP
#    define setEncPinInput(x) setPinInputHigh(x)
#  elif ENCODER_PIN_MODE == ENC_PDOWN
#    define setEncPinInput(x) setPinInputLow(x)
#  elif ENCODER_PIN_MODE == ENC_FLOAT
#    define setEncPinInput(x) setPinInput(x)
#  endif
#else
#  define setEncPinInput(x) setPinInputHigh(x)
#endif

    for (int i = 0; i < NUMBER_OF_ENCODERS; i++) {
        setEncPinInput(encoders_pad_a[i]);
        setEncPinInput(encoders_pad_b[i]);

        encoder_state[i] = (readPin(encoders_pad_a[i]) << 0) | (readPin(encoders_pad_b[i]) << 1);
    }
```

Comment (drashna) [reply quantum/encoder.c:112]: Could even include: 
```c
#if defined(ENCODER_PIN_PDOWN)
#    ifdef __AVR__
#        Error Pull Down is not supported on AVR
#    endif 
#    define setEncPinInput(x) setPinInputLow(x)
#elif defined(ENCODER_PIN_FLOAT)
#    define setEncPinInput(x) setPinInput(x)
#elif defined(ENCODER_PIN_OUTPUT
#    define setEncPinInput(x) setPinOutput(x)
#else
#    define setEncPinInput(x) setPinInputHigh(x)
#endif
```

Not sure if output would even work, though.

**Comment by drashna** [comment]
Also, ideally, this should probably be targeting develop.

**Comment by Croktopus** [comment]
alright so i basically combined your suggestions and it looks much cleaner now, thank you both. how can i redirect it to develop, or should i close this and open a new one?

**Comment by Croktopus** [comment]
> alright so i basically combined your suggestions and it looks much cleaner now, thank you both. how can i redirect it to develop, or should i close this and open a new one?

Ok slight problem it doesnt work now lol. getting on discord

**Comment by drashna** [comment]
To redirect it to develop, edit the title.  That should let you choose what branch to target.

**Review by drashna — APPROVED** [review]

**Review by mtei — APPROVED** [review]

File: quantum/encoder.c [hunk 3]
```diff
@@ -81,12 +81,33 @@ void encoder_init(void) {
...
 
-        encoder_state[i] = (readPin(encoders_pad_a[i]) << 0) | (readPin(encoders_pad_b[i]) << 1);
-    }
+for (int i = 0; i < NUMBER_OF_ENCODERS; i++) {
+    setEncPinInput(encoders_pad_a[i]);
+    setEncPinInput(encoders_pad_b[i]);
+
+    encoder_state[i] = (readPin(encoders_pad_a[i]) << 0) | (readPin(encoders_pad_b[i]) << 1);
+}
```
Comment (zvecr) [inline quantum/encoder.c:110]: clang-format
```suggestion
    for (int i = 0; i < NUMBER_OF_ENCODERS; i++) {
        setEncPinInput(encoders_pad_a[i]);
        setEncPinInput(encoders_pad_b[i]);

        encoder_state[i] = (readPin(encoders_pad_a[i]) << 0) | (readPin(encoders_pad_b[i]) << 1);
    }
```

File: quantum/encoder.c [hunk 4]
```diff
@@ -81,12 +81,33 @@ void encoder_init(void) {
     }
 #endif
 
-    for (int i = 0; i < NUMBER_OF_ENCODERS; i++) {
-        setPinInputHigh(encoders_pad_a[i]);
-        setPinInputHigh(encoders_pad_b[i]);
+#ifdef ENCODER_PINS
```
Comment (zvecr) [inline quantum/encoder.c:84]: Naming things is difficult, however i'm cautious of too much crossover to existing setting, eg `BACKLIGHT_PINS`. While docs exists, it leads to questions like "why doesn't `#define ENCODER_PINS {A1, B2}` work?"

I cant find an existing setting to align to, but the closest thing that come to mind would be `ENCODER_PIN_MODE`, maybe `ENCODER_PAD_MODE`.

**Review by zvecr — CHANGES_REQUESTED** [review]

**Comment by tzarc** [comment]
Sorry about that, GitHub decided to delete the `develop` branch from upstream when we merged it, despite being told otherwise. Reopened.

**Pull request closed without merging**